guidance/models/transformers/_transformers.py

```diff
@@ -174,6 +77,204 @@ def __init__(
+                        encoded_str = transformers_tokenizer.encode(token_str)
+                        if len(encoded_str) != 1:
+                            raise ValueError(f"Round-trip encoding of tokens [{token}] failed! Got {encoded_str}")
+                        roundtrip_id = encoded_str[0]
+                        if roundtrip_id == i:
+                            byte_coded = token_str.encode()
+                        else:
+                            byte_coded = bytes([byte_decoder[c] for c in token])
+                    else:
+                        byte_coded = token.encode()
+                else:
+                    raise ValueError(f"Unexpected token type: {type(token)}")
+            byte_tokens[i] = byte_coded
+        return byte_tokens
+
+    def _byte_tokens_fallback(
+        self,
+        transformers_tokenizer: Union[
+            "transformers_package.PreTrainedTokenizer",
+            "transformers_package.PreTrainedTokenizerFast",
+        ],
+    ) -> list[bytes]:
+        byte_tokens = [b""] * len(transformers_tokenizer)
+        byte_decoder: dict[str, int] = transformers_package.AutoTokenizer.from_pretrained(
+            "gpt2", use_fast=False
+        ).byte_decoder  # fall back to gpt2 mapping
+
+        # some special tokens may not have their whitespace encoded...
+        byte_decoder[" "] = 32
+        byte_decoder["\n"] = 10
+        byte_decoder["\r"] = 13
+        byte_decoder["\t"] = 9
+        byte_decoder["▁"] = 32
+
+        # run a quick spot check to verify we can rebuild complex multi-token unicode symbols
+        s = "’•¶∂ƒ˙∆£Ħ爨ൠᅘ∰፨"
+        reconstructed = b""
+        try:
+            input_ids = transformers_tokenizer(s)["input_ids"]
+            for i in input_ids:
+                nxt_bytes = []
+                token_str = transformers_tokenizer.convert_ids_to_tokens(i)
+                for c in token_str:
+                    nxt_bytes.append(byte_decoder[c])
+                reconstructed += bytes(nxt_bytes)
+            # Check if the tokenizer has a bos_token attribute, and if it does, check
+            # if it's at the start of the reconstructed bytes
+            # Some tokenizers add this automatically as part of the call function, so
+            # we need to remove it to compare
+            if hasattr(transformers_tokenizer, "bos_token") and reconstructed.startswith(
+                transformers_tokenizer.bos_token.encode()
+            ):
+                reconstructed = reconstructed[len(transformers_tokenizer.bos_token) :]
+        except Exception as e:
+            msg = textwrap.dedent(
+                f"""
+                The tokenizer being used is unable to convert a special character in {s}.
+                For models with sentencepiece based tokenizers (e.g. llama, phi-3-mini),
+                installing sentencepiece often fixes this issue (pip install sentencepiece).
+                """
+            )
+            raise ValueError(msg) from e
```

Comment:
I'm unsure of whether this method will ever be called now that we "fail early" with respect to the fast tokenizer. If we hit this exception, the message may be wrong now, as missing an import probably isn't the cause of the exception anymore...